File: src/Site/Hellraiser.php
<?php

declare(strict_types=1);

namespace MetalLE\Site;

use MetalLE\Event\Event;
use MetalLE\Event\Location;

class Hellraiser implements Site
{
    private const DATE_SEARCH = [
        'Mo, ',
        'Di, ',
        'Mi, ',
        'Do, ',
        'Fr, ',
        'Sa, ',
        'So, ',
        ' Januar ',
        ' Februar ',
        ' März ',
        ' April ',
        ' Mai ',
        ' Juni ',
        ' Juli ',
        ' August ',
        ' September ',
        ' Oktober ',
        ' November ',
        ' Dezember ',
    ];
    private const DATE_REPLACE = [
        '',
        '',
        '',
        '',
        '',
        '',
        '',
        '01.',
        '02.',
        '03.',
        '04.',
        '05.',
        '06.',
        '07.',
        '08.',
        '09.',
        '10.',
        '11.',
        '12.',
    ];

    public function getIterator(): \Traversable
    {
        $location = new Location(
            'hr',
            'Hellraiser Leipzig',
            'https://hellraiser-leipzig.de/',
            'img/hellraiser.png',
        );

        yield from $this->parseEvents($location,'https://hellraiser-leipzig.de/produkt-kategorie/tickets/');
        yield from $this->parseEvents($location,'https://hellraiser-leipzig.de/produkt-kategorie/tickets/page/2/');
        yield from $this->parseEvents($location,'https://hellraiser-leipzig.de/produkt-kategorie/tickets/page/3/');
    }

    /**
     * @return Event[]
     */
    private function parseEvents(Location $location, string $url): iterable
    {
        $plainHTML = file_get_contents($url);
        if (preg_match_all(
                '#<li class="product.*<a href="(.*)".*<img.*src="(.*)".*<h2.*>(.*)</h2>.*<div class="date-published">(.*)</div>.*</li>#isU',
                $plainHTML,
                $matches,
                PREG_SET_ORDER
            ) > 0) {
            foreach ($matches as $match) {
                $name = str_replace(['Ticket &#8222;', '&#8220;'], '', $match[3]);
                $name = html_entity_decode($name, ENT_HTML5);
                $date = str_replace(self::DATE_SEARCH, self::DATE_REPLACE, $match[4]);

                $event = new Event($name, new \DateTimeImmutable($date), $location, $match[1], $match[2]);
                yield $event->getID() => $event;
            }
        }
    }
}

File: src/Site/hellraiser.rs
use std::borrow::Borrow;

use html_escape::decode_html_entities;
use regex::Regex;

use crate::tools::date::parse_german_date;
use crate::{Event, Location, Site, HTTP};

const URL: &str = "https://hellraiser-leipzig.de";

pub struct Hellraiser {
    location: Location,
}

impl Hellraiser {
    pub(crate) fn new() -> Self {
        Self {
            location: Location {
                slug: "hr".to_string(),
                name: "Hellraiser Leipzig".to_string(),
                website: URL.to_string(),
            },
        }
    }
}

impl Site for Hellraiser {
    fn get_locations(&self) -> Vec<Location> {
        return vec![self.location.clone()];
    }

    fn fetch_events(&self) -> Vec<Event> {
        let http = HTTP::new();

        let mut result = Vec::new();

        let reg = Regex::new("(?si)<li class=\"product.*?<a href=\"(?P<url>.*?)\".*?<img.*?src=\"(?P<img>.*?)\".*?<h2.*?>(?P<name>.*?)</h2>.*?<div class=\"date-published\">.*?, (?P<date>.*?)</div>.*?</li>").unwrap();

        for i in 1..10 {
            let html =
                http.get(&(URL.to_string() + "/produkt-kategorie/tickets/page/" + &*i.to_string()));

            for capture in reg.captures_iter(html.as_str()) {
                let name = capture.name("name").unwrap().as_str().to_string();
                let name = name.replace("Ticket &#8222;", "");
                let name = name.replace("&#8220;", "");

                let evt = Event::new(
                    decode_html_entities(name.as_str()).to_string(),
                    parse_german_date(capture.name("date").unwrap().as_str()).and_hms(0, 0, 0),
                    self.location.borrow(),
                    capture.name("url").unwrap().as_str().to_string(),
                    Some(capture.name("img").unwrap().as_str().to_string()),
                );

                result.push(evt);
            }

            if !html.contains("class=\"next page-numbers\"") {
                break;
            }
        }

        result
    }
}
